pkg/trace/remoteconfighandler/remote_config_handler.go

```diff
@@ -0,0 +1,179 @@
+type RemoteConfigHandler struct {
+	remoteClient          config.RemoteClient
+	prioritySampler       prioritySampler
+	errorsSampler         errorsSampler
+	rareSampler           rareSampler
+	agentConfig           *config.AgentConfig
+	remoteConfigPathRegex *regexp.Regexp
+}
+
+func New(conf *config.AgentConfig, prioritySampler prioritySampler, rareSampler rareSampler, errorsSampler errorsSampler) *RemoteConfigHandler {
+	if conf.RemoteSamplingClient == nil {
+		return nil
+	}
+
+	// paths are in the form datadog/<org ID>/<product name>/<config ID>/config
+	r, err := regexp.Compile(`datadog\/\d+\/APM_SAMPLING\/(.+)\/config`)
+	if err != nil {
+		panic(err)
+	}
+
+	return &RemoteConfigHandler{
+		remoteClient:          conf.RemoteSamplingClient,
+		prioritySampler:       prioritySampler,
+		rareSampler:           rareSampler,
+		errorsSampler:         errorsSampler,
+		agentConfig:           conf,
+		remoteConfigPathRegex: r,
+	}
+}
+
+func (h *RemoteConfigHandler) Start() {
+	if h == nil {
+		return
+	}
+
+	h.remoteClient.Start()
+	h.remoteClient.RegisterAPMUpdate(h.onUpdate)
+}
+
+func (h *RemoteConfigHandler) onUpdate(update map[string]state.APMSamplingConfig) {
+	remoteRateUpdates, samplerconfigPayload, err := h.extractUpdatePayloads(update)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if len(remoteRateUpdates) != 0 {
+		h.prioritySampler.UpdateRemoteRates(remoteRateUpdates)
+	}
+
+	if samplerconfigPayload != nil {
+		h.updateSamplers(*samplerconfigPayload)
+	}
+}
+
+func (h *RemoteConfigHandler) extractUpdatePayloads(update map[string]state.APMSamplingConfig) ([]sampler.RemoteRateUpdate, *apmsampling.SamplerConfig, error) {
+	var remoteRateUpdates []sampler.RemoteRateUpdate
+	var samplerconfigPayload *apmsampling.SamplerConfig
+	for path, conf := range update {
+		configID, err := h.configPathToID(path)
+		if err != nil {
+			return []sampler.RemoteRateUpdate{}, nil, err
+		}
+
+		if configID == dynamicRatesByService || configID == userDefinedRatesByService {
+			var payload apmsampling.APMSampling
+			_, err := payload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal APMSampling remote config envPayload: %w", err)
+			}
+			remoteRateUpdates = append(remoteRateUpdates, sampler.RemoteRateUpdate{Version: conf.Metadata.Version, Config: payload})
+		}
+		if configID == samplerconfig {
+			samplerconfigPayload = &apmsampling.SamplerConfig{}
+			_, err := samplerconfigPayload.UnmarshalMsg(conf.Config)
+			if err != nil {
+				return []sampler.RemoteRateUpdate{}, nil, fmt.Errorf("failed to unmarshal SamplerConfig remote config envPayload: %w", err)
+			}
+		}
+	}
+	return remoteRateUpdates, samplerconfigPayload, nil
+}
+
+func (h *RemoteConfigHandler) configPathToID(path string) (apmsamplingRemoteConfigID, error) {
+	pathMatch := h.remoteConfigPathRegex.FindStringSubmatch(path)
```

Comment:
Probably not a big deal since I don't think this will be a hot path, but something to keep in mind is that regexp in golang can be a bit slow. It would probably be faster to just split on the `/` and look for the pieces we need. I don't think this needs to change

Reply:
Thanks for the tip, I'll keep that in mind :)